pkg/resources/openstack/network.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"strings"
+
+	osrouter "github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+
+	"k8s.io/kops/pkg/resources"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+const (
+	typeRouterIF = "Router-IF"
+	typeRouter   = "Router"
+	typeSubnet   = "Subnet"
+	typeNetwork  = "Network"
+)
+
+func (os *clusterDiscoveryOS) ListNetwork() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+
+	opt := networks.ListOpts{
+		Name: os.clusterName,
+	}
+	networks, err := os.osCloud.ListNetworks(opt)
+	if err != nil {
+		return resourceTrackers, err
+	}
+
+	for _, network := range networks {
+		optRouter := osrouter.ListOpts{
+			Name: strings.Replace(os.clusterName, ".", "-", -1),
```

Comment:
There's no reference to network in the router logic...  Accidental?  Hoist outside the loop?

Reply:
hmm true, that could be outside `networks` for loop. Now it is executed too many times